core/trino-main/src/main/java/io/trino/execution/scheduler/UniformNodeSelector.java

```diff
@@ -168,47 +158,37 @@ public SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTas
         queueSizeAdjuster.update(existingTasks, assignmentStats);
         Set<InternalNode> blockedExactNodes = new HashSet<>();
         boolean splitWaitingForAnyNode = false;
-        // splitsToBeRedistributed becomes true only when splits go through locality-based assignment
-        boolean splitsToBeRedistributed = false;
-        Set<Split> remainingSplits = new HashSet<>(splits.size());
 
         List<InternalNode> filteredNodes = filterNodes(nodeMap, includeCoordinator, ImmutableSet.of());
         ResettableRandomizedIterator<InternalNode> randomCandidates = new ResettableRandomizedIterator<>(filteredNodes);
         Set<InternalNode> schedulableNodes = new HashSet<>(filteredNodes);
 
-        // optimizedLocalScheduling enables prioritized assignment of splits to local nodes when splits contain locality information
-        if (optimizedLocalScheduling) {
-            for (Split split : splits) {
-                if (split.isRemotelyAccessible() && !split.getAddresses().isEmpty()) {
-                    List<InternalNode> candidateNodes = selectExactNodes(nodeMap, split.getAddresses(), includeCoordinator);
-
-                    Optional<InternalNode> chosenNode = candidateNodes.stream()
-                            .filter(ownerNode -> assignmentStats.getTotalSplitsWeight(ownerNode) < maxSplitsWeightPerNode && assignmentStats.getUnacknowledgedSplitCountForStage(ownerNode) < maxUnacknowledgedSplitsPerTask)
-                            .min(comparingLong(assignmentStats::getTotalSplitsWeight));
-
-                    if (chosenNode.isPresent()) {
-                        assignment.put(chosenNode.get(), split);
-                        assignmentStats.addAssignedSplit(chosenNode.get(), split.getSplitWeight());
-                        splitsToBeRedistributed = true;
-                        continue;
-                    }
-                }
-                remainingSplits.add(split);
-            }
-        }
-        else {
-            remainingSplits = splits;
-        }
-
-        for (Split split : remainingSplits) {
+        for (Split split : splits) {
             randomCandidates.reset();
 
             List<InternalNode> candidateNodes;
+            boolean exactNodes;
             if (!split.isRemotelyAccessible()) {
                 candidateNodes = selectExactNodes(nodeMap, split.getAddresses(), includeCoordinator);
+                exactNodes = true;
             }
             else {
-                candidateNodes = selectNodes(minCandidates, randomCandidates);
+                // optimizedLocalScheduling enables prioritized assignment of splits to local nodes when splits contain locality information
+                if (optimizedLocalScheduling && !split.getAddresses().isEmpty()) {
+                    candidateNodes = selectExactNodes(nodeMap, split.getAddresses(), includeCoordinator);
```

Comment:
I think we can also need to change implementation of `hive.force-local-scheduling`. With current changes we're effectively forcing local scheduling by default in hive for HDFS deployments (unless node went down). That's probably not the right default as it could create hot spots in the cluster. Instead of flipping `isRemotelyAccessible`, that flag should now change whether or not host addresses are returned by hive connector and by default we shouldn't provide addresses from HDFS.

Reply:
Apart from `UniformNodeSelector` and FTE there is also `TopologyAwareNodeSelector` which was not changed. In case of `TopologyAwareNodeSelector` setting `hive.force-local-scheduling` changes the policy rather considerably and `ConnectorSplit` doesn't provide any hard contract on split node affinity (just hint to schedule policy).

> by default we shouldn't provide addresses from HDFS.

I think it's OK as long as HDFS returns at least few addresses (which I think it does by default). Previously, `addresses` were also honored by default, just less strict. Therefore disabling `address` propagation would probably be bigger change. However, we still have `optimizedLocalScheduling` for the rescue if something goes bad.

Reply:
With `UniformNodeSelector` (the most common case), the flag `hive.force-local-scheduling` basically becomes useless now. You get forced local scheduling either way and the only way to opt out is to turn off `node-scheduler.optimized-local-scheduling` completely which affects all catalogs. This doesn't look like a good state to be in.
> Previously, addresses were also honored by default, just less strict
There is a big difference in how hard the scheduler will try to use provided addresses now. It's basically equal to `isRemotelyAccessible = false` from before (modulo the corner case of node going away).
> ConnectorSplit doesn't provide any hard contract on split node affinity (just hint to scheduling policy).
I no longer care what the `ConnectorSplit` java doc says. It is not being updated to reflect the actual reality that the contract has changed. Node addresses are now no longer just "hint", they are strictly enforced unless the node goes down. This is in fact the same scheduling policy as `isRemotelyAccessible = false` (modulo the corner case of node going away).